server/src/internalClusterTest/java/org/opensearch/gateway/remote/RemoteRoutingTableServiceIT.java

```diff
@@ -0,0 +1,271 @@
+        ensureStableCluster(4);
+
+        assertBusy(() -> {
+            int indexRoutingFilesAfterNodeDrop = repository.blobStore().blobContainer(indexRoutingPath).listBlobs().size();
+            assertTrue(indexRoutingFilesAfterNodeDrop > indexRoutingFiles.get());
+        });
+
+        RemoteClusterStateService remoteClusterStateService = internalCluster().getClusterManagerNodeInstance(
+            RemoteClusterStateService.class
+        );
+        RemoteManifestManager remoteManifestManager = remoteClusterStateService.getRemoteManifestManager();
+        verifyUpdatesInManifestFile(remoteManifestManager);
+    }
+
+    public void testRemoteRoutingTableIndexMasterRestart() throws Exception {
+        BlobStoreRepository repository = prepareClusterAndVerifyRepository();
+
+        List<RoutingTable> routingTableVersions = getRoutingTableFromAllNodes();
+        assertTrue(areRoutingTablesSame(routingTableVersions));
+
+        // Ensure node comes healthy after restart
+        String clusterManagerName = internalCluster().getClusterManagerName();
+        internalCluster().restartNode(clusterManagerName);
+        ensureGreen();
+        ensureGreen(INDEX_NAME);
+
+        // ensure master is elected and the cluster is stable
+        assertNotNull(internalCluster().clusterService().state().nodes().getClusterManagerNode());
+        ensureStableCluster(4);
+
+        assertBusy(() -> {
+            int indexRoutingFilesAfterNodeDrop = repository.blobStore().blobContainer(indexRoutingPath).listBlobs().size();
+            assertTrue(indexRoutingFilesAfterNodeDrop > indexRoutingFiles.get());
+        });
+
+        RemoteClusterStateService remoteClusterStateService = internalCluster().getClusterManagerNodeInstance(
+            RemoteClusterStateService.class
+        );
+        RemoteManifestManager remoteManifestManager = remoteClusterStateService.getRemoteManifestManager();
+        verifyUpdatesInManifestFile(remoteManifestManager);
+    }
+
+    private BlobStoreRepository prepareClusterAndVerifyRepository() throws Exception {
+        prepareCluster(1, 3, INDEX_NAME, 1, 5);
+        ensureGreen(INDEX_NAME);
+
+        RepositoriesService repositoriesService = internalCluster().getClusterManagerNodeInstance(RepositoriesService.class);
+        BlobStoreRepository repository = (BlobStoreRepository) repositoriesService.repository(REPOSITORY_3_NAME);
+
+        BlobPath baseMetadataPath = getBaseMetadataPath(repository);
+        List<IndexRoutingTable> indexRoutingTables = new ArrayList<>(getClusterState().routingTable().indicesRouting().values());
+        indexRoutingPath = getIndexRoutingPath(
+            baseMetadataPath.add(INDEX_ROUTING_PATH_TOKEN),
+            indexRoutingTables.get(0).getIndex().getUUID()
+        );
+
+        assertBusy(() -> {
+            indexRoutingFiles.set(repository.blobStore().blobContainer(indexRoutingPath).listBlobs().size());
+            assertEquals(6, indexRoutingFiles.get());
```

Comment:
Can we run 100 iterations of the test and ensure this verification is not flaky?

Reply:
Tested it across run, count of files was not same all the time.

Reply:
Lets keep a deterministic logic only which will be true everytime.